lib/src/util/console_log_utils.dart

```diff
@@ -152,3 +154,51 @@ class ConsoleConfig {
   /// Captures calls to `console.log`, `console.warn`, and `console.error`.
   static const ConsoleConfig all = ConsoleConfig._('all');
 }
+
+final formatRegExp = RegExp('%[sdj%]');
+String format(dynamic f, List<dynamic> arguments) {
+  if (f is! String) {
+    final objects = [];
+    for (var i = 0; i < arguments.length; i++) {
+      objects.add(arguments[i]);
+    }
+    return objects.join(' ');
+  }
+
+  var str = '';
+  if (f is String) {
+    if (!f.contains('%')) return f;
+
+    var i = 0;
+    final args = arguments;
+    final len = args.length;
+    str += f.replaceAllMapped(formatRegExp, (m) {
+      final x = m[0].toString();
+      if (x == '%%') return '%';
+      if (i >= len) return x;
+      switch (x) {
+        case '%s':
+          return args[i++].toString();
+        case '%d':
+          return num.tryParse(args[i++].toString()).toString();
+        case '%j':
+          try {
+            return callMethod(getProperty(window,'JSON'), 'stringify', [args[i++]]).toString();
```

Comment:
Could we use static JS interop here instead?


```dart
@JS('JSON.stringify')
external String _jsonStringify(dynamic object);
```

```suggestion
            return _jsonStringify(args[i++]);
```

Reply:
ugh but that would mean i have to like make this a library, give it a name, and add package js, ugh im exhausted already